libcaf_core/caf/async/promise.hpp

```diff
@@ -27,8 +27,8 @@ class promise {
   }
 
   promise& operator=(const promise& other) noexcept {
-    promise copy{other};
-    cell_.swap(copy.cell_);
+    if (this != &other)
+      return *this = promise{other};
```

Comment:
The default constructor for `promise` increments the number of promises by 1, and omitting it here breaks the unit tests, so we can't just do it like suggested. Either keep the copy-swap or add the missing line for `fetch_add`, and I like this approach more because of DRY.